File: content/en/04-llms/llms/_demos/16_demo_tools/16_weather-tool-app.R
library(shiny)
library(bslib)
library(ellmer)
library(shinychat)
library(weathR)

get_weather <- tool(
  \(lat, lon) {
    # Tools must return a string, JSON, or Content object.
    # ellmer 0.5.0 no longer auto-converts data frames.
    forecast <- sf::st_drop_geometry(weathR::point_forecast(lat, lon))
    jsonlite::toJSON(forecast, auto_unbox = TRUE)
  },
  name = "get_weather",
  description = "Get forecast data for a specific latitude and longitude.",
  arguments = list(
    lat = type_number("Latitude of the location."),
    lon = type_number("Longitude of the location.")
  )
)

ui <- page_fillable(
  chat_ui("chat")
)

server <- function(input, output, session) {
  client <- ellmer::chat_posit(model = "zai-org/GLM-5.3-Flash")
  client$register_tool(get_weather)

  chat_server("chat", client)
}

shinyApp(ui, server)

File: content/en/04-llms/llms/_demos/16_demo_tools/16_weather-tool-app.py
# ---- ✦ Weather tool, in a Shiny app ✦ ----
import NWS

import chatlas
from shiny import App, ui
from shinychat import Chat, chat_ui

NWS.InitiateAPI("posit::conf(2026)", "conf@posit.co")


def get_weather(lat: float, lon: float):
    """
    Get forecast data for a specific latitude and longitude.

    Parameters
    ----------
    lat : str
        Latitude of the location.
    lon : str
        Longitude of the location.
    """
    return NWS.GetCurrentForecast(lat, lon)


app_ui = ui.page_fillable(chat_ui("chat"))


def server(input, output, session):
    chat_client = chatlas.ChatPosit(model="zai-org/GLM-5.3-Flash")
    chat_client.register_tool(get_weather)

    Chat("chat", client=chat_client)


app = App(app_ui, server)
